File: leetcode/1500-1600/1594. 矩阵的最大非负积.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
MOD = int(1e9 + 7)


class Solution:
    def maxProductPath(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        MOD = 10**9 + 7

        # dp[i][j][0] 存储最大值，dp[i][j][1] 存储最小值
        dp = [[[-inf, inf] for _ in range(n)] for _ in range(m)]

        # 初始化起始点
        dp[0][0] = [grid[0][0], grid[0][0]]

        for j in range(1, n):
            x = grid[0][j]
            candidates = [x * dp[0][j - 1][0], x * dp[0][j - 1][1]]
            dp[0][j][0] = max(candidates)
            dp[0][j][1] = min(candidates)

        # 正确的第一列初始化
        for i in range(1, m):
            x = grid[i][0]
            candidates = [x * dp[i - 1][0][0], x * dp[i - 1][0][1]]
            dp[i][0][0] = max(candidates)
            dp[i][0][1] = min(candidates)

        # 填充剩余位置
        for i in range(1, m):
            for j in range(1, n):
                x = grid[i][j]
                candidates = [
                    x * dp[i - 1][j][0],  # 从上方来的最大值
                    x * dp[i - 1][j][1],  # 从上方来的最小值
                    x * dp[i][j - 1][0],  # 从左方来的最大值
                    x * dp[i][j - 1][1],  # 从左方来的最小值
                ]
                dp[i][j][0] = max(candidates)
                dp[i][j][1] = min(candidates)

        result = dp[m - 1][n - 1][0]
        return result % MOD if result >= 0 else -1

File: leetcode/1500-1600/1594. 矩阵的最大非负积.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <functional>
#include <algorithm>
#include <ranges>
#include <numeric>
#include <unordered_set>
#include <cmath>   
#include <memory> 
#include <map>
#include <queue>
#include <cstring>

using namespace std;

class Solution {
public:
    int maxProductPath(vector<vector<int>> &grid) {
        int n = grid.size();
        if (n == 0) return -1; // Handle empty grid case
        int m = grid[0].size();
        if (m == 0) return -1; // Handle empty row case

        long long mod = 1e9 + 7;

        // dp[i][j].first = max product to reach (i, j)
        // dp[i][j].second = min product to reach (i, j)
        // Using 0-based indexing consistent with grid
        vector<vector<pair<long long, long long>>> dp(n, vector<pair<long long, long long>>(m));

        // Initialize the starting cell
        dp[0][0].first = grid[0][0];
        dp[0][0].second = grid[0][0];

        // Initialize the first row (can only come from the left)
        for (int j = 1; j < m; ++j) {
            long long curr = grid[0][j];
            long long prev_max = dp[0][j - 1].first;
            long long prev_min = dp[0][j - 1].second;
            // Max and min are determined by multiplying the current value with the previous max and min
            dp[0][j].first = max(prev_max * curr, prev_min * curr);
            dp[0][j].second = min(prev_max * curr, prev_min * curr);
        }

        // Initialize the first column (can only come from the top)
        for (int i = 1; i < n; ++i) {
            long long curr = grid[i][0];
            long long prev_max = dp[i - 1][0].first;
            long long prev_min = dp[i - 1][0].second;
            dp[i][0].first = max(prev_max * curr, prev_min * curr);
            dp[i][0].second = min(prev_max * curr, prev_min * curr);
        }

        // Fill the rest of the dp table
        for (int i = 1; i < n; ++i) {
            for (int j = 1; j < m; ++j) {
                long long curr = grid[i][j];

                // Get max and min products from the cell above
                long long up_max = dp[i - 1][j].first;
                long long up_min = dp[i - 1][j].second;

                // Get max and min products from the cell to the left
                long long left_max = dp[i][j - 1].first;
                long long left_min = dp[i][j - 1].second;

                // Calculate the four possible product candidates
                long long prod1 = up_max * curr;
                long long prod2 = up_min * curr;
                long long prod3 = left_max * curr;
                long long prod4 = left_min * curr;

                // The new max is the maximum of all candidates
                dp[i][j].first = max({ prod1, prod2, prod3, prod4 });
                // The new min is the minimum of all candidates
                dp[i][j].second = min({ prod1, prod2, prod3, prod4 });
            }
        }

        // Get the maximum product at the bottom-right cell
        long long final_max_prod = dp[n - 1][m - 1].first;

        // If the maximum product is negative, return -1
        if (final_max_prod < 0) {
            return -1;
        } else {
            // Otherwise, return the result modulo 10^9 + 7
            // Ensure the result of modulo is non-negative if needed, though final_max_prod is guaranteed non-negative here.
            return (int)(final_max_prod % mod);
        }
    }
};